### ISSUE TITLE
zenity fails to start when file chooser's path has an apostrophe

### ISSUE THREAD

**Issue by junetried** [fltk/fltk#884]
**Describe the bug**
When opening a native file chooser from a directory where the path to it has an apostrophe (`'`), zenity will fail to start. Instead, an error like this will appear in the app's output:

```
sh: -c: line 1: syntax error near unexpected token `('
sh: -c: line 1: `zenity  --title 'Pick a file' --file-selection --filename='/tmp/directory with a ' in the name/'  --file-filter='Text (*.txt)|*.txt' --file-filter='C Files (*.cxx *.h *.c *.cpp)|*.cxx *.h *.c *.cpp' --file-filter='Tars (*.tar *.tar.gz)|*.tar *.tar.gz' --file-filter='Apps (*.app)|*.app' --file-filter='All files (*)|*' 2> /dev/null'
```

and the selection will be considered aborted. Choosing a directory fails with a different error:

```
sh: -c: line 1: unexpected EOF while looking for matching `''
```

**To Reproduce**
In the `native-filechooser` example, enter any path with an apostrophe and try to open it.

**Expected behavior**
The file chooser should open like normal.

**FLTK Version**
 - Version: `fltk-config --version` says 1.3.8, but the latest tag is 1.3.9 and I cloned the git repo
 - Downloaded and self-built from: Git
 - If from Git, branch: master
 - If from Git, commit: 2db1929985f430d3be1f87222aa85599bb2b8222

**FLTK Configure / Build Options**
`cmake -D OPTION_USE_WAYLAND=0` as I don't have something installed and don't really want to figure out what

**Operating System / Platform:**
 - OS: Linux
 - OS Version: openSUSE Tumbleweed snapshot 20231228

**Linux/Unix Runtime, if applicable:**
X11

**Comment by Albrecht-S** [comment]
Thanks for the report. Using an apostrophe in a filename is unusual but legal, hence it seems like it would be best to escape it when creating the zenity command line.

I'm marking this as bug but I'm not going to work on it. Hopefully another dev will have a look at this and fix it.

@ManoloFLTK I believe the zenity specific code was added by you, would you like to fix it? Are there other commandline driven file choosers (GTK file chooser, maybe) that could suffer from the same issue?

**Comment by Albrecht-S** [comment]
@junetried For your information:

Please take care not to mix FLTK versions. It appears that you have a system version (`1.3.8`) installed, so if you execute `fltk-config` it uses the installed version from your PATH.

In our Git repo the latest **_release_** is indeed `1.3.9` (tag: `release-1.3.9`) but the `master` branch is `1.4.0` (not yet released, hence no Git tag). The zenity filechooser is not available in 1.3.x and will not be added.

If you build FLTK yourself from the master branch or a current snapshot (see download page) you'll get the development version `1.4.0` (to be released soon). The feature to use the zenity filechooser has been added in 1.4.0 (Git master, current development).

**Comment by junetried** [comment]
> @junetried For your information:
> 
> Please take care not to mix FLTK versions. It appears that you have a system version (`1.3.8`) installed, so if you execute `fltk-config` it uses the installed version from your PATH.
> 
> In our Git repo the latest **_release_** is indeed `1.3.9` (tag: `release-1.3.9`) but the `master` branch is `1.4.0` (not yet released, hence no Git tag). The zenity filechooser is not available in 1.3.x and will not be added.
> 
> If you build FLTK yourself from the master branch or a current snapshot (see download page) you'll get the development version `1.4.0` (to be released soon). The feature to use the zenity filechooser has been added in 1.4.0 (Git master, current development).

Good catch! I forgot to run the `fltk-config` in the build directory instead of the system version. Oops.

**Comment by erco77** [comment]
> @ManoloFLTK I believe the zenity specific code was added by you, would you like to fix it?
> Are there other commandline driven file choosers (GTK file chooser, maybe) that could suffer from the same issue?

I gave this a look briefly, and it's probably due to our use of popen() in Fl_Native_File_Chooser_Kdialog.cxx, which invokes a shell (/bin/sh) to run the command, so arguments containing shell-special characters (like ; & | quotes, etc) can all run afoul with the shell interpreter that popen() invokes.

You can try to escape all those characters, but it gets messy, and is generally "dangerous".

IMHO the right solution would instead be to implement our own popen()-like function which uses execv() or execvp() so that one can build an argv[] array (avoiding building a string with whitespace delimited arguments) so that the command can be invoked directly without an intermediate shell, avoiding shell-special characters messing things up.

This would involve fork() and wait() and dup() and pipe() for stdio redirection, which gets tricky if you're not familiar with it, but if done right can be clean.

So unless Manolo is familiar with that stuff too, I might be the most qualified to address this, as I'm all too familiar with that stuff.

That's not an easy fix, as all the command building would have to transition from building a string dynamically, to building an argv[] array dynamically, so I don't think this can be squeezed into 1.4.0, but surely in a maintenance release.

This popen() alternative would probably be a useful FLTK library "internal only" class, because popen() is somewhat insecure due to the shell interpreter; it's too easy for execution injections, so refitting other use of popen() throughout fltk might be useful.

**Comment by Albrecht-S** [comment]
> That's not an easy fix, as all the command building would have to transition from building a string dynamically, to building an argv[] array dynamically, so I don't think this can be squeezed into 1.4.0, but surely in a maintenance release.

... or in 1.5.0 when we can use std::string for all kinds of dynamic strings and string "editing" if necessary. Maybe.

However, I don't know what we should do meanwhile. Is there a simple "quick and dirty" fix for the most important stuff? Even if we only quoted the apostrophe (single quote)?

Just a thought (w/o looking at the code at all): if we quoted all commandline arguments with single quotes, would other special shell characters like `'&'` matter **inside** the quoted strings? Probably not, I think.

**Comment by erco77** [comment]
A quick and dirty would be to replace all single quotes with literally  `'"'"'` which breaks down to:

    <close-single-quote> <open-double-quote> <single-quote> <close-double-quote> <open-single-quote>

With that, the filename: `don't_open_me.txt`
..would become (on the shell command line): `'don'"'"'t_open_me.txt'`

There's other ways, equally ugly but involves backslashes, which IMHO cause more trouble, as they themselves have to be escaped for every parser they pass through (c compiler, shell, etc).

**Comment by erco77** [comment]
I'll take a stab at doing the quoted-quote technique with the new Fl_String.

**Comment by erco77** [comment]
Attaching a patch suggestion here as [issue_844_patch_v1.txt](https://github.com/fltk/fltk/files/13897812/issue_844_patch_v1.txt)

I'd apply it, but I don't have a way to test the "kdialog" code, as I don't (and can't currently) have KDE, and not sure how to force it to be used. (If I rename away the zenity executable, it seems to fall back to using the plugin.. not sure where that logic is offhand)

Tested and works for zenity: test/native-filechooser lets you pick and re-pick files with single quotes.
So in the case of a two files named:

    foo'bar.txt
    foo'bar'bla".txt

The command line for the first ends up being:

    zenity --title 'Pick a file' --file-selection --filename='/usr/local/src/fltk-1.4.x.git/build/foo'"'"'bar.txt' --file-filter='Text (*.txt)|*.txt' --file-filter='C Files (*.cxx *.h *.c *.cpp)|*.cxx *.h *.c *.cpp' --file-filter='Tars (*.tar *.tar.gz)|*.tar *.tar.gz' --file-filter='Apps (*.app)|*.app' --file-filter='All files (*)|*' 2> /dev/null

..the key bit being:

   --filename='/usr/local/src/fltk-1.4.x.git/build/foo'"'"'bar.txt' 

..and the key bit for the second file ends up being:

    --filename='/usr/local/src/fltk-1.4.x.git/build/foo'"'"'bar'"'"'bla".txt'

..and both seem to work.

I left the debugging messages enabled that print to stdout the command passed to popen() in the patch.
I used the same quoting algorithm for some of the other command line arguments that were dynamic.

The quoting algorithm (the new `shell_quote()` method) does not do anything special for utf8 -- I'm assuming (but can't remember) if utf8 multibyte encoding of Unicode /doesn't/ include single quotes within them. IIRC the ASCII set is excluded from utf8 multibyte's, right?

Anyway, it's a start -- would appreciate it if someone with KDE could test the kdialog code, and/or give the patch a once over.

**Comment by erco77** [comment]
@junetried -- if you want to try the patch above, be sure to pull the current 1.4.x from git and checkout commit dc70a0410 before applying the patch; the patch depends on that commit.

**Comment by ManoloFLTK** [comment]
@erco77 I've tested your patch with KDE and `kdialog`: it works OK with apostrophe in filename or in directory name or in both. Thanks.

**Comment by erco77** [comment]
@ManoloFLTK - thanks; some questions:

- Is there a way I can force the "kdialog" command to run, even if the executable isn't installed?
_i.e. by tweaking the FNFC src code? I just want to carefully exercise the command construction in the various cases and check the debug msgs._
- Is it OK to use Fl_String for method parameters as I have here?
_I know we're trying to keep Fl_String out of public facing headers. So I'm assuming the kdialog/zenity code I've modified here is all private because the .H files are all in the src dir, thereby not public facing._

I'll remove the debugging and apply the patch later today.

**Comment by ManoloFLTK** [comment]
> @ManoloFLTK - thanks; some questions:
> 
>     * Is there a way I can force the "kdialog" command to run, even if the executable isn't installed?
>       _i.e. by tweaking the FNFC src code? I just want to carefully exercise the command construction in the various cases and check the debug msgs._

You'd have to force env var `XDG_CURRENT_DESKTOP` to value `KDE` and to arrange for command `kdialog -v` to run and output some text (any text is OK) to stdout (see lines 951ff of `Fl_Native_File_Chooser_GTK.cxx`). 
With this, FLTK will run `kdialog` to create file chooser windows.


> 
>     * Is it OK to use Fl_String for method parameters as I have here?
>       _I know we're trying to keep Fl_String out of public facing headers. So I'm assuming the kdialog/zenity code I've modified here is all private because the .H files are all in the src dir, thereby not public facing._

Yes, that is completely OK because none of classes `Fl_Kdialog_Native_File_Chooser_Driver` and `Fl_Zenity_Native_File_Chooser_Driver` are public.

**Comment by ManoloFLTK** [comment]
and also, you'd need to turn Zenity off via `fltk-options `as shown below:
<img width="652" alt="image" src="https://github.com/fltk/fltk/assets/41016272/da723b02-02ca-4139-804f-dd8593f3184a">

**Comment by Albrecht-S** [comment]
A tiny issue with the patch (unused variable):
```
Building CXX object src/CMakeFiles/fltk.dir/Fl_Native_File_Chooser_Zenity.cxx.o
/git/fltk/master/src/Fl_Native_File_Chooser_Zenity.cxx: In member function ‘virtual void Fl_Zenity_Native_File_Chooser_Driver::build_command(Fl_String&)’:
/git/fltk/master/src/Fl_Native_File_Chooser_Zenity.cxx:81:7: warning: unused variable ‘l’ [-Wunused-variable]
   81 |   int l;
      |       ^
```

**Comment by Albrecht-S** [comment]
FTR: I tried to compile after replacing `Fl_String` with `std::string`, as in the following patch (on top of yours), and it worked well with another minor issue (if/when `USE_STD` was enabled):
```diff
diff --git a/src/Fl_Native_File_Chooser_Kdialog.H b/src/Fl_Native_File_Chooser_Kdialog.H
index 8b50b36fa..b4e9d15ea 100644
--- a/src/Fl_Native_File_Chooser_Kdialog.H
+++ b/src/Fl_Native_File_Chooser_Kdialog.H
@@ -25,5 +25,12 @@
 
 #include <FL/Fl_Native_File_Chooser.H>
+
+#if (FLTK_USE_STD)
+#include <string>
+typedef std::string Fl_String;
+#else
 #include "Fl_String.H"
+#endif
+
 
 class Fl_Kdialog_Native_File_Chooser_Driver : public Fl_Native_File_Chooser_FLTK_Driver {
diff --git a/src/Fl_Native_File_Chooser_Kdialog.cxx b/src/Fl_Native_File_Chooser_Kdialog.cxx
index 9ec829a67..1fa322c98 100644
--- a/src/Fl_Native_File_Chooser_Kdialog.cxx
+++ b/src/Fl_Native_File_Chooser_Kdialog.cxx
@@ -26,5 +26,4 @@
 #include "Fl_Window_Driver.H"
 #include "drivers/Unix/Fl_Unix_Screen_Driver.H"
-#include "Fl_String.H"
 #include <stdio.h>
 #include <stdlib.h>
```

Notes:
1. It turned out that including `Fl_String.H` in `src/Fl_Native_File_Chooser_Kdialog.cxx` is not necessary and caused a conflict after my changes in `src/Fl_Native_File_Chooser_Kdialog.H`. You can remove this `#include` statement; this will simplify edits in 1.5.0 when we remove `Fl_String`.
2. Another compilation warning `[-Wsign-compare]` was issued because `Fl_String::size()` returns `int` whereas `std::string::size()` returns unsigned long. We should not care about thisnow, it's easy to fix in 1.5.
3. This was only an experiment. We don't need to take care of these minor issues in 1.4.x, they can be addressed in 1.5.0.

@erco77 Final note: I didn't test anything except building with both options, leaving the runtime tests to you and @ManoloFLTK.

Thanks for providing and testing this patch.

**Comment by erco77** [comment]
OK, applied as d9125d76c.

@ManoloFLTK Great - thanks for the info and test.
@Albrecht-S  Thanks for the audit - I removed the unused 'int l' and the unnecessary #include.
